Lib/tokenize.py

```diff
@@ -492,8 +492,11 @@ def _tokenize(readline, encoding):
             # BOM will already have been stripped.
             encoding = "utf-8"
         yield TokenInfo(ENCODING, encoding, (0, 0), (0, 0), '')
+    last_line = b''
+    line = b''
     while True:                                # loop over lines in stream
         try:
+            last_line = line
```

Comment:
Shouldn't `last_line` only be set after `StopIteration` is caught? ISTM that in other cases we wouldn't want to be adding the newline at the end.

Reply:
`readline` is one of the ancient APIs that existed before generators. There's two ways of stopping iteration, either raising `StopIteration` or returning the [empty string](https://docs.python.org/3.7/tutorial/inputoutput.html#methods-of-file-objects), the latter gets caught all the way down here https://github.com/python/cpython/blob/master/Lib/tokenize.py#L528

How you're describing it is what I had initially but there's a few places in the loop where iteration can stop so I thought this would be simpler.

Reply:
Then perhaps add a comment to that end?  It seems rather crucial to understanding that particular piece of the code.

Reply:
Good point, added.